README.md

```diff
@@ -1,2 +1,103 @@
+       git checkout develop
+ 
+ * Once cloned, run a maven build either from your IDE or use below command:
+
+        mvn clean install
+        java -jar target/cost-report-service-0.0.1-SNAPSHOT.jar 
+ 
+ * Once application is started type the url in a browser:
+ 
+      http://localhost:8080/swagger-ui.html
+      
+ * NOTE* Make sure no other application is using port 8080.
+ 
+ ## Toolset:
+  1) JDK 11.0.3
+  2) Spring Boot v2.4.3
+  4) spring-boot-starter-data-jpa v2.4.3 (Spring data JPA defaults to hibernate)
+  5) springfox-swagger-ui/springfox-swagger2 v2.9.4
+  
+ ## APIs
+   * POST /api/report/production/{production}/ep (process single EP cost report)
+   * GET api/report/production/{production}?epCode={epCode} (get aggregated cost report given production name and EP number)
+   * POST /api/report/production/{production} (process production cost report)
+   * GET api/report/production/{production} (get aggregated production cost report)
+   * POST /api/report/production/{production}/amortized (process production cost report with amortized costs)
+   * GET api/report/production/{production} (get aggregated final production cost report including amortized costs)
+
+**NOTE:** I strongly recommend using swagger-ui to see the complete API documentation.
+
+ ## Error Scenarios and Exception Handling:
+
+ Customized InvalidProductionException and InvalidReportException are created to log all application exceptions generated
+ by the service during request handling.
+ Error scenarios includes:
+  1) get aggregated cost report by invalid production name (404 status code)
+  2) get cost report by invalid production/EP name combination (404 status code)
+  3) post request with empty {production} path param in URL (400 status code)
+  4) single EP Cost input payload contains more than one epCode (400 status code)
+  5) production cost input payload is empty (400 status code)
+
+**NOTE:** I did not consider negative amount to be error case, considering we may have some refund transactions.
+
+
+ ## Database:
+
+Java In-memory h2 database is used. During start up table create scripts are run and sample data set inserted.
+
+ ### DataModel:
+ 
+ The datamodel has 1 table:
+  1) EpTotalCostEntity (store aggregated cost on production EP level)
+
+    long id;    (auto-generated id as primary key)
+  	String prodName;  (production name, could be used as foreign key to join with proudction table to get production details)
+  	String epCode;  (epCode)
+  	double itemizedTotalCost;   (aggregated itemized total cost)
+  	double amortizedCost;   (calculated amortized cost for this proudction and epCode)
+  	double totalCost;   (itemizedTotalCost + amortizedCost)
+  	long timestamp; (used for version control. we should always fetch the cost with latest timestamp)
+  
+ Potentially we could add one more table to record each itemized cost from input data for historical data retrieval purpose:
```

Comment:
I think we assumed that the records would be stored at the line-item grain but didn't make that explicit in the docs.  It doesn't make a difference to this particular problem, however.

During our pair session let's discuss tradeoffs between computing the report at ingestion, at query time, or possibly out-of-band.

Reply:
Thanks for the comments. Yes, the requirement document does not mention the need to store each itemized cost. But I thought it's a good idea to keep trace of everything. 
Regarding when to compute the final aggregated cost report, my opinion is that we should not do it on the fly, means during client query time, especially if the final cost will be retrieved again and again. Rest of two options are potentially solutions, and we should decide based on data scale as well sync/async requirement from clients.